onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -145,18 +166,18 @@ bool ReshapeFusion::Fuse_Subgraph1(Node& reshape, Graph& graph, const logging::L
   shape_initializer_proto.set_data_type(ONNX_NAMESPACE::TensorProto_DataType_INT64);
   shape_initializer_proto.set_raw_data(shape_value.data(), shape_value.size() * sizeof(int64_t));
   auto& new_node_arg = graph_utils::AddInitializer(graph, shape_initializer_proto);
-  if (!graph_utils::ReplaceNodeWithInitializer(graph, *graph.GetNode(concat.Index()), new_node_arg)) {
-    return false;
-  }
 
-  // Remove nodes that are not used anymore.
-  for (const auto& node_type_and_index : candidates_for_removal) {
-    Node* node = graph.GetNode(node_type_and_index.second);
-    if (node->GetOutputEdgesCount() == 0 && graph.GetNodeOutputsInGraphOutputs(*node).empty()) {
-      graph.RemoveNode(node->Index());
+  // Safely remove concat parent nodes which have only one output
+  for (int i = 0; i < concat_input_count; ++i) {
+    const Node* p_cur_node = graph_utils::GetInputNode(concat, i);
+    if (p_cur_node != nullptr) {
+      graph_utils::RemoveNodesWithOneOutputBottomUp(graph, p_cur_node);
     }
   }
 
+  if (!graph_utils::ReplaceNodeWithInitializer(graph, *graph.GetNode(concat.Index()), new_node_arg)) {
```

Comment:
In what situation could this returns false?
If so, this is better to place before removing nodes in case this failed.

Reply:
I think this one was from the original code. CanReplaceNodeWithInitializer(.) is called in line 158.